File: prototype/base/src/main/java/com/project/base/service/JwtService.java
package com.project.base.service;

import com.project.base.config.JwtProperties;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.time.Instant;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class JwtService {

    private final JwtProperties props;
    private final SecretKey key;

    public JwtService(JwtProperties props) {
        this.props = props;
        this.key = Keys.hmacShaKeyFor(props.getSecret().getBytes());
    }

    public String generate(String username, List<String> roles) {
        Instant now = Instant.now();
        Instant exp = now.plusSeconds(props.getAccessTokenTtlSeconds());

        return Jwts.builder()
                .setSubject(username)
                .setIssuer(props.getIssuer())
                .setIssuedAt(Date.from(now))
                .setExpiration(Date.from(exp))
                .claim("roles", roles)
                .signWith(key)
                .compact();
    }

    public String parseUsername(String token) {
        return Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token)
                .getBody()
                .getSubject();
    }

    public List<String> parseRoles(String token) {
        List<?> rawRoles = (List<?>) Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token)
                .getBody()
                .get("roles");

        return rawRoles.stream()
                .map(Object::toString)
                .collect(Collectors.toList());
    }
}

File: prototype/lietzsche/src/main/kotlin/com/stxtory/lietzsche/service/JwtService.kt
package com.stxtory.lietzsche.service

import com.stxtory.lietzsche.config.JwtProperties
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.*

@Service
class JwtService(private val props: JwtProperties) {

    private val key = Keys.hmacShaKeyFor(props.secret.toByteArray())

    fun generate(username: String, roles: List<String>): String {
        val now = Instant.now()
        val exp = now.plusSeconds(props.accessTokenTtlSeconds)
        return Jwts.builder()
            .setSubject(username)
            .setIssuer(props.issuer)
            .setIssuedAt(Date.from(now))
            .setExpiration(Date.from(exp))
            .claim("roles", roles)
            .signWith(key)
            .compact()
    }

    fun parseUsername(token: String): String =
        Jwts.parserBuilder().setSigningKey(key).build()
            .parseClaimsJws(token).body.subject

    fun parseRoles(token: String): List<String> =
        (Jwts.parserBuilder().setSigningKey(key).build()
            .parseClaimsJws(token).body["roles"] as List<*>)
            .map { it.toString() }
}